tensorflow/compiler/xla/service/gpu/gemm_thunk.h

```diff
@@ -79,9 +102,49 @@ Status RunGemm(
     se::DeviceMemoryBase rhs_buffer, se::DeviceMemoryBase output_buffer,
     se::Stream* stream, bool implements_whole_instruction,
     absl::optional<int64_t> profile_index,
+    BlasScratchAllocator* scratch_allocator,
+    se::blas::IBlasLtMatmulAlgorithm* const profiled_algorithm,
     se::blas::ProfileResult* profile_result = nullptr,
     absl::optional<se::blas::AlgorithmType> algorithm = absl::nullopt);
 
+using MatrixDescs =
+    std::tuple<se::blas::MatrixDescriptor, se::blas::MatrixDescriptor,
+               se::blas::MatrixDescriptor>;
+MatrixDescs PopulateInputOutputMatrices(const GpuGemmConfig& gemm_config,
+                                        se::DeviceMemoryBase lhs_buffer,
+                                        se::DeviceMemoryBase rhs_buffer,
+                                        se::DeviceMemoryBase output_buffer);
+
+inline bool BlasPlansCompatibleType(PrimitiveType type) {
+  absl::flat_hash_set<PrimitiveType> compatible_types = {F16, F32, F64, C64, C128};
+  return compatible_types.contains(type);
+}
+
+// A class for storing and retrieving algorithms in cublasLT autotuning
+class BlasPlansAutotuneCache {
+ public:
+  BlasPlansAutotuneCache() {}
+  bool Find(const se::BatchMatmulParameters& params,
+            se::blas::AlgorithmConfig* config);
+  void Insert(const se::BatchMatmulParameters& params,
+              const se::blas::AlgorithmConfig& config);
+
+ private:
+  struct Hasher {
+    std::size_t operator()(const se::BatchMatmulParameters& parameter) const {
+      return parameter.hash();
+    }
+  };
+
+  mutable absl::Mutex mu_;
+  absl::flat_hash_map<se::BatchMatmulParameters, se::blas::AlgorithmConfig,
```

Comment:
Usually types are made hashable like this: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/hlo_value.h;l=59-62?q=AbslHashValue%20file:xla